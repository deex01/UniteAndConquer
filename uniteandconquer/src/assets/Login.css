/* CSS file for Login.js */
 
.logo {
  text-align: center;
  font-size: 25px;
}
 
h2 {
  font-size: 25px;
}
 
input.field {
  border: none;
  background-color: #FFD9A0;
  margin: 5px;
  width: 60%;
  height: 30px;
  padding: 10px;
  font-size: 25px;
}

input:hover {
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
}
 
::placeholder {
  font-style: italic;
  font-size: 25px;
}
 
.login {
  width: 500px;
  margin: auto;
  text-align: center;
  border: 10px solid #FFD9A0;
  box-sizing: border-box;
}
 
.submit {
  border: none;
  background: #FFCB7D;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 5px;
  padding: 15px;
  width: 60%;
 
  font-style: italic;
  font-weight: bold;
  font-size: 25px;
}
 
.submit:hover {
  box-shadow: 5px 5px 5px rgba(0, 0, 0, 0.25);
  background-color: grey;
  transition-duration: 0.4s;
}
