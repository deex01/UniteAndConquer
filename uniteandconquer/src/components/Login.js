/* eslint-disable no-unused-vars */
import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import '../assets/Login.css';
import '../assets/App.css';

const UserDB = require('../modules/UserDB');

function Login() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const navigate = useNavigate();
  const myStorage = window.sessionStorage;
  // We need to ensure that the username and password are not null
  // And the username and password are in good format, if any.
  // And compare whether the username and password match.

  // We check whether a user using email or phone number as username
  // by checking whether the username_ include '@'.
  const checkPassword = (username_, password_) => {
    if (username_.includes('@')) {
      UserDB.loginUserWithEmail(username_, password_, (success, id, err) => {
        if (success) {
          myStorage.setItem('UserID', id);
        } else {
          console.log(err);
        }
        return success;
      });
    } else {
      UserDB.loginUserWithPhone(username_, password_, (success, id, err) => {
        if (success) {
          myStorage.setItem('UserID', id);
        } else {
          console.log(err);
        }
        return success;
      });
    }
  };
  const login = () => {
    if (username.length <= 0 || password.length <= 0) {
      throw new Error('Invalid username and password. It cannot be empty');
    } else if (!checkPassword(username, password)) {
      throw new Error('Incorrect password or username not exists');
    } else if (checkPassword(username, password)) {
      navigate('/');
    }
  };

  return (
    <div className="login-page">
      <div className="logo"><h1>Unite and Conquer</h1></div>
      <div className="login">
        <h2>Login to your account</h2>
        <div className="login-field">
          <input data-testid="email/phone-input" className="field" placeholder="email/phone #" onChange={(e) => setUsername(e.target.value)} />
        </div>
        <div className="login-field">
          <input data-testid="password-input" className="field" placeholder="password" onChange={(e) => setPassword(e.target.value)} />
        </div>
        <br />
        <button className="submit" type="button" onClick={login}>
          login
        </button>
        <p>
          Do not already have an account?
          {' '}
          <Link to="/registration">Sign up</Link>
          {' '}
        </p>
      </div>

    </div>
  );
}

export default Login;
