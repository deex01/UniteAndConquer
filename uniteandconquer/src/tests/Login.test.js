/**
 * @jest-environment jsdom
 */

import React from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import {
  render, screen, fireEvent,
} from '@testing-library/react';
import Login from '../components/Login';

// https://stackoverflow.com/questions/66284286/react-jest-mock-usenavigate

const mockedUsedLink = jest.fn();
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: () => mockedUsedLink,
  useNavigate: () => mockedUsedNavigate,
}));

test('login header', () => {
  render(<Login />);
  const headerElement = screen.getByText('Login to your account');
  expect(headerElement).toBeInTheDocument();
});

// first field

test('input email', () => {
  render(<Login />);
  const input = screen.getByTestId('email/phone-input');
  fireEvent.change(input, { target: { value: 'john@gmail.com' } });
  expect(input.value).toBe('john@gmail.com');
});

test('input phone number', () => {
  render(<Login />);
  const input = screen.getByTestId('email/phone-input');
  fireEvent.change(input, { target: { value: '234-234-2345' } });
  expect(input.value).toBe('234-234-2345');
});

// password field

test('input valid password', () => {
  render(<Login />);
  const input = screen.getByTestId('password-input');
  fireEvent.change(input, { target: { value: 'Aa3#bcad' } });
  expect(input.value).toBe('Aa3#bcad');
});

// test('log in with email', () => {
//   render(<Login />);
//   try {
//     fireEvent.click(screen.getByText('login'));
//   } catch (e) {
//     expect(e.message).toBe('Invalid username and password. It cannot be empty');
//   }
// });

// test('log in with email', () => {
//   render(<Login />);
//   const user = screen.getByTestId('email/phone-input');
//   fireEvent.change(user, { target: { value: 'john@gmail.com' } });
//   expect(user.value).toBe('john@gmail.com');
//   const input = screen.getByTestId('password-input');
//   fireEvent.change(input, { target: { value: 'Aa3#bcad' } });
//   expect(input.value).toBe('Aa3#bcad');
//   expect(() => fireEvent.click(screen.getByText('login')).
// toThrow('Incorrect password or username not exists'));
// });

test('snapshot registration', () => {
  const component = renderer.create(<Login />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
