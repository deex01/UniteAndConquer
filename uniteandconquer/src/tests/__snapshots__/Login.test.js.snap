// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot registration 1`] = `
<div
  className="login-page"
>
  <div
    className="logo"
  >
    <h1>
      Unite and Conquer
    </h1>
  </div>
  <div
    className="login"
  >
    <h2>
      Login to your account
    </h2>
    <div
      className="login-field"
    >
      <input
        className="field"
        data-testid="email/phone-input"
        onChange={[Function]}
        placeholder="email/phone #"
      />
    </div>
    <div
      className="login-field"
    >
      <input
        className="field"
        data-testid="password-input"
        onChange={[Function]}
        placeholder="password"
      />
    </div>
    <br />
    <button
      className="submit"
      onClick={[Function]}
      type="button"
    >
      login
    </button>
    <p>
      Do not already have an account?
       
       
    </p>
  </div>
</div>
`;
